dbms/src/Functions/arrayEnumerateRanked.h

```diff
@@ -0,0 +1,358 @@
+
+    ColumnArray::Offset prev_off = 0;
+
+    using Map = ClearableHashMap<
+        UInt128,
+        UInt32,
+        UInt128TrivialHash,
+        HashTableGrower<INITIAL_SIZE_DEGREE>,
+        HashTableAllocatorWithStackMemory<(1ULL << INITIAL_SIZE_DEGREE) * sizeof(UInt128)>>;
+    Map indices;
+
+    /*
+
+(2, [[1,2,3],[2,2,1],[3]], 2, [4,5,6], 1)
+    ; 1 2 3;  2 2 1;  3        4 5 6
+    ; 4 4 4;  5 5 5;  6      <-
+
+(1, [[1,2,3],[2,2,1],[3]], 1, [4,5,6], 1)
+    ;[1,2,3] [2,2,1] [3]       4 5 6
+    ;4       5       6       <-
+
+(1, [[1,2,3],[2,2,1],[3]], 1, [4,5,6], 0)
+    ;[1,2,3] [2,2,1] [3]       4 5 6
+    ;[4,5,6] [4,5,6] [4,5,6] <-
+
+. - get data
+; - clean index
+
+(1, [[[1,2,3],[1,2,3],[1,2,3]],[[1,2,3],[1,2,3],[1,2,3]],[[1,2]]], 1)
+    ;.                         .                         .
+
+(1, [[[1,2,3],[1,2,3],[1,2,3]],[[1,2,3],[1,2,3],[1,2,3]],[[1,2]]], 2)
+    ; .       .       .         .       .       .         .
+
+(2, [[[1,2,3],[1,2,3],[1,2,3]],[[1,2,3],[1,2,3],[1,2,3]],[[1,2]]], 2)
+    ; .       .       .       ; .       .       .       ; .
+
+(1, [[[1,2,3],[1,2,3],[1,2,3]],[[1,2,3],[1,2,3],[1,2,3]],[[1,2]]], 3)
+    ;  . . .   . . .   . . .     . . .   . . .   . . .     . .
+
+(2, [[[1,2,3],[1,2,3],[1,2,3]],[[1,2,3],[1,2,3],[1,2,3]],[[1,2]]], 3)
+    ;  . . .   . . .   . . .  ;  . . .   . . .   . . .  ;  . .
+
+(3, [[[1,2,3],[1,2,3],[1,2,3]],[[1,2,3],[1,2,3],[1,2,3]],[[1,2]]], 3)
+    ;  . . . ; . . . ; . . .  ;  . . . ; . . . ; . . .  ;  . .
+
+*/
+
+    std::vector<size_t> indexes_by_depth(max_array_depth);
+    std::vector<size_t> current_offset_n_by_depth(max_array_depth);
+
+    UInt32 rank = 0;
+
+    for (size_t off : offsets)
+    {
+        std::vector<size_t> indexes(columns.size());
+
+        bool want_clear = false;
+
+        for (size_t j = prev_off; j < off; ++j)
+        {
+            for (size_t col_n = 0; col_n < columns.size(); ++col_n)
+                indexes[col_n] = indexes_by_depth[depths[col_n] - 1];
+
+            auto hash = hash128depths(indexes, columns);
+
+            if constexpr (std::is_same_v<Derived, FunctionArrayEnumerateUniqRanked>)
+            {
+                auto idx = ++indices[hash];
+                res_values[j] = idx;
+            }
+            else // FunctionArrayEnumerateDenseRanked
+            {
+                auto idx = indices[hash];
+                if (!idx)
+                {
+                    idx = ++rank;
+                    indices[hash] = idx;
+                }
+                res_values[j] = idx;
+            }
+
+            // Debug: DUMP(off, prev_off, j, indexes, res_values[j], columns);
+
+            for (int depth = current_offset_depth - 1; depth >= 0; --depth)
+            {
+                ++indexes_by_depth[depth];
+
+                if (indexes_by_depth[depth] == (*offsets_by_depth[depth])[current_offset_n_by_depth[depth]])
+                {
+                    if (static_cast<int>(clear_depth - 1) == depth)
```

Comment:
Can check that `depth + 1 == clear_depth` to avoid static_cast.

Reply:
../dbms/src/Functions/arrayEnumerateRanked.h:336:37: error: comparison of integers of different signs: 'DB::DepthType' (aka 'unsigned int') and 'int' [-Werror,-Wsign-compare]
                    if (clear_depth == depth + 1)
                        ~~~~~~~~~~~ ^  ~~~~~~~~~